lib/Analysis/SecretnessAnalysis/SecretnessAnalysis.cpp

```diff
@@ -109,26 +109,85 @@ LogicalResult SecretnessAnalysis::visitOperation(
 }
 
 void annotateSecretness(Operation *top, DataFlowSolver *solver) {
+  // Attribute "Printing" Helper
+  auto getAttribute =
+      [&](const SecretnessLattice *secretnessLattice) -> NamedAttribute {
+    if (!secretnessLattice) {
+      return {StringAttr::get(top->getContext(), "secretness is `null`"),
+              UnitAttr::get(top->getContext())};
+    }
+    if (!secretnessLattice->getValue().isInitialized()) {
+      return {StringAttr::get(top->getContext(), "secretness is `unknown`"),
+              UnitAttr::get(top->getContext())};
+    }
+    if (secretnessLattice->getValue().getSecretness()) {
+      return {StringAttr::get(top->getContext(),
+                              secret::SecretDialect::kArgSecretAttrName),
+              UnitAttr::get(top->getContext())};
+    }
+    return {StringAttr::get(top->getContext(), "public"),
+            UnitAttr::get(top->getContext())};
+  };
+
   // Add an attribute to the operations to show determined secretness
   top->walk([&](Operation *op) {
-    for (unsigned i = 0; i < op->getNumResults(); ++i) {
-      std::string name = op->getNumResults() == 1
-                             ? "secretness"
-                             : "result_" + std::to_string(i) + "_secretness";
-      auto *secretnessLattice =
-          solver->lookupState<SecretnessLattice>(op->getOpResult(i));
-      if (!secretnessLattice) {
-        op->setAttr(name, StringAttr::get(op->getContext(), "null"));
-        return;
+    // Custom Handling for `func.func`, which uses special attributes
+    if (auto func = llvm::dyn_cast<func::FuncOp>(op)) {
+      // The arguments are either already typed as !secret.secret<..>
+      // or annotated with {secret.secret}, so we don't need to update
+      // them However, the result is generally not, so we need to
+      // annotate that.
+
+      auto *ret = func.getFunctionBody().back().getTerminator();
+      assert(ret->getNumOperands() == func.getFunctionType().getNumResults() &&
+             "Number of returned values does not match function type");
+      for (unsigned i = 0; i < func.getFunctionType().getNumResults(); ++i) {
+        auto *secretnessLattice =
+            solver->lookupState<SecretnessLattice>(ret->getOpOperand(i).get());
+        // We cannot simple add our normal StringAttr or BoolAttr, as
+        // `func.func` only allows dialect attributes as annotations for
+        // arguments and results, so we must use the {secret.secret}
+        // style:
+        if (secretnessLattice &&
+            secretnessLattice->getValue().isInitialized() &&
+            secretnessLattice->getValue().getSecretness()) {
+          func.setResultAttr(
+              i,
+              StringAttr::get(func->getContext(),
+                              secret::SecretDialect::kArgSecretAttrName),
+              UnitAttr::get(func->getContext()));
```

Comment:
```suggestion
          func.setResultAttr(
              i,
              secret::SecretDialect::kArgSecretAttrName,
              UnitAttr::get(func->getContext()));
```

setResultAttr has an overload that accepts a StringRef, so you shouldn't need to wrap it in a StringAttr.

https://github.com/llvm/llvm-project/blob/a976036a100b7dd459b6cabac96159875fcd513d/mlir/include/mlir/Interfaces/FunctionInterfaces.td#L530